File: scripts/extract-diac.js
// Extrae placas y tapacantos PVC del JSON DIAC → CSVs editables.
//
// Input:  C:/Users/emara/Desktop/productos-diac.json
// Output: scripts/lista-placas.csv
//         scripts/lista-tapacantos-pvc.csv
//
// Filtro de tapacantos: SOLO PVC (excluye aluminio, anodizados, productos químicos).
// Parser tolerante: si no logra extraer un campo, lo deja vacío. El usuario edita.

const fs = require("fs");
const path = require("path");

const INPUT = "C:/Users/emara/Desktop/productos-diac.json";
const OUT_DIR = __dirname;

// ── Detectores ──────────────────────────────────────────────────────────
// Orden importa: más específico primero.
const TIPO_DETECTORS = [
  [/\bench(?:apad[oa]|\.)\b/i, "enchapado"],
  [/\bmel(?:amina|am)?\b/i,    "melamina"],
  [/\bmdf\b/i,                 "mdf"],
  [/\baglo(?:merado)?\b/i,     "aglomerado"],
  [/\bterc(?:iado)?\b/i,       "terciado"],
  [/\bmac(?:iza)?\b/i,         "madera_maciza"],
  [/\bmfc\b/i,                 "melamina"],
  [/\bgua(?:tambu)?\b/i,       "terciado"],
  [/\bfibrofacil\b/i,          "mdf"],
];

const ESPESOR_RE  = /(\d+(?:\.\d+)?)\s*mm/i;
const DIMS_RE     = /(\d\.\d{2})\s*[xX/]\s*(\d\.\d{2})\s*m/i;
const DIMS_ALT_RE = /(\d{3,4})\s*[xX]\s*(\d{3,4})/i;

// Tapacanto PVC: incluye "Tapa Canto" + "PVC" o "ABS" o "melamínico p/canto" o "PVC Cubrec".
// Excluye "A.mte", "anod" (aluminio anodizado).
const TAPACANTO_PVC_RE = /\b(pvc\s*(?:cubrec|tapacanto|canto)|abs\s*canto|melam[íi]nico\s*p\.?\s*canto|tapacanto\s*pvc)\b/i;
const ES_ALUMINIO_RE = /\b(a\.\s*mte|aluminio|anod|alum\.)\b/i;

function detectarTipo(desc) {
  for (const [rx, tipo] of TIPO_DETECTORS) {
    if (rx.test(desc)) return tipo;
  }
  return "";
}

function extraerEspesor(desc) {
  const m = desc.match(ESPESOR_RE);
  return m ? parseFloat(m[1]) : "";
}

function extraerDims(desc) {
  let m = desc.match(DIMS_RE);
  if (m) return [Math.round(parseFloat(m[1]) * 1000), Math.round(parseFloat(m[2]) * 1000)];
  m = desc.match(DIMS_ALT_RE);
  if (m) {
    const a = parseInt(m[1], 10), b = parseInt(m[2], 10);
    if (a >= 1000 && b >= 1000) return [a, b];
  }
  return ["", ""];
}

function limpiarNombre(desc) {
  let s = desc;
  s = s.replace(new RegExp(ESPESOR_RE.source, "gi"), "");
  s = s.replace(new RegExp(DIMS_RE.source,    "gi"), "");
  s = s.replace(new RegExp(DIMS_ALT_RE.source,"gi"), "");
  for (const [rx] of TIPO_DETECTORS) {
    s = s.replace(new RegExp(rx.source, "gi"), "");
  }
  s = s.replace(/\b(euca|euda|rec\s*\w+|simil|ln|x\d{2,}|lc|lex?\d+|xa\d+)\b/gi, "");
  s = s.replace(/\s+/g, " ").replace(/^[\s\-·.,]+|[\s\-·.,]+$/g, "");
  return s;
}

function esPlaca(desc) {
  if (!detectarTipo(desc)) return false;
  if (extraerEspesor(desc) === "") return false;
  const [a, b] = extraerDims(desc);
  return a !== "" && b !== "";
}

function esTapacantoPVC(desc) {
  if (ES_ALUMINIO_RE.test(desc)) return false;
  return TAPACANTO_PVC_RE.test(desc);
}

function toCsvField(v) {
  const s = String(v == null ? "" : v);
  if (/[",\n]/.test(s)) return `"${s.replace(/"/g, '""')}"`;
  return s;
}

function rowsToCsv(rows, headers) {
  const out = [headers.map(toCsvField).join(",")];
  for (const r of rows) {
    out.push(headers.map(h => toCsvField(r[h])).join(","));
  }
  return out.join("\n");
}

// ── Run ─────────────────────────────────────────────────────────────────
console.log(`Leyendo ${INPUT}...`);
const productos = JSON.parse(fs.readFileSync(INPUT, "utf-8"));
console.log(`Total productos: ${productos.length}`);

const placas = [];
const tapacantos = [];
let descartados = 0;

for (const p of productos) {
  const desc = p.descripcion || "";
  const codigoDiac = p.codigo_diac || "";
  const precio = p.precio || "";

  if (esTapacantoPVC(desc)) {
    tapacantos.push({
      codigo:           "",
      nombre:           desc,
      espesor:          extraerEspesor(desc) || "",
      precio,
      codigoDiac,
      descripcionDiac:  desc,
    });
    continue;
  }

  if (esPlaca(desc)) {
    const tipo = detectarTipo(desc);
    const espesor = extraerEspesor(desc);
    const [largo, ancho] = extraerDims(desc);
    placas.push({
      codigo:           "",
      nombre:           limpiarNombre(desc),
      tipo,
      espesor,
      placaLargo:       largo,
      placaAncho:       ancho,
      precioPlaca:      precio,
      codigoDiac,
      descripcionDiac:  desc,
    });
    continue;
  }

  descartados += 1;
}

placas.sort((a, b) =>
  (a.tipo || "").localeCompare(b.tipo || "") ||
  (a.espesor || 0) - (b.espesor || 0) ||
  (a.nombre || "").localeCompare(b.nombre || "")
);
tapacantos.sort((a, b) => (a.nombre || "").localeCompare(b.nombre || ""));

const placasCsv = rowsToCsv(placas, ["codigo", "nombre", "tipo", "espesor", "placaLargo", "placaAncho", "precioPlaca", "codigoDiac", "descripcionDiac"]);
const tapaCsv   = rowsToCsv(tapacantos, ["codigo", "nombre", "espesor", "precio", "codigoDiac", "descripcionDiac"]);

const PLACAS_PATH = path.join(OUT_DIR, "lista-placas.csv");
const TAPA_PATH   = path.join(OUT_DIR, "lista-tapacantos-pvc.csv");

fs.writeFileSync(PLACAS_PATH, placasCsv, "utf-8");
fs.writeFileSync(TAPA_PATH,   tapaCsv,   "utf-8");

console.log(`\nResultados:`);
console.log(`  Placas extraidas:        ${placas.length}`);
console.log(`  Tapacantos PVC:          ${tapacantos.length}`);
console.log(`  Descartados:             ${descartados}`);

// Distribución por tipo (placas)
const porTipo = {};
for (const p of placas) porTipo[p.tipo] = (porTipo[p.tipo] || 0) + 1;
console.log(`\nPlacas por tipo:`);
for (const [t, n] of Object.entries(porTipo).sort((a,b) => b[1] - a[1])) {
  console.log(`  ${t.padEnd(15)} ${n}`);
}

console.log(`\nArchivos generados:`);
console.log(`  ${PLACAS_PATH}`);
console.log(`  ${TAPA_PATH}`);

File: scripts/extract-diac.py
"""
Extrae placas y tapacantos del JSON DIAC y genera CSVs editables.

Input:  C:/Users/emara/Desktop/productos-diac.json
Output: scripts/lista-placas.csv
        scripts/lista-tapacantos.csv

Cada CSV tiene columnas para que el usuario complete/ajuste y la app importe.
Parser tolerante: si no logra extraer un campo, lo deja vacio.
"""
import json
import csv
import re
import os
from pathlib import Path

INPUT = Path(r"C:/Users/emara/Desktop/productos-diac.json")
OUT_DIR = Path(__file__).parent

# ── Detectores ─────────────────────────────────────────────────────────────
# Orden importa: mas especifico primero.
TIPO_DETECTORS = [
    (re.compile(r"\bench(?:apad[oa]|\.)\b", re.IGNORECASE), "enchapado"),
    (re.compile(r"\bmel(?:amina|am)?\b",     re.IGNORECASE), "melamina"),
    (re.compile(r"\bmdf\b",                  re.IGNORECASE), "mdf"),
    (re.compile(r"\baglo(?:merado)?\b",      re.IGNORECASE), "aglomerado"),
    (re.compile(r"\bterc(?:iado)?\b",        re.IGNORECASE), "terciado"),
    (re.compile(r"\bmac(?:iza)?\b",          re.IGNORECASE), "madera_maciza"),
    (re.compile(r"\bmfc\b",                  re.IGNORECASE), "melamina"),
    (re.compile(r"\bgua(?:tambu)?\b",        re.IGNORECASE), "terciado"),
    (re.compile(r"\bfibrofacil\b",           re.IGNORECASE), "mdf"),
]

ESPESOR_RE = re.compile(r"(\d+(?:\.\d+)?)\s*mm", re.IGNORECASE)
DIMS_RE = re.compile(r"(\d\.\d{2})\s*[xX/]\s*(\d\.\d{2})\s*m", re.IGNORECASE)
DIMS_ALT_RE = re.compile(r"(\d{3,4})\s*[xX]\s*(\d{3,4})", re.IGNORECASE)

TAPACANTO_RE = re.compile(r"\b(tapa\s*canto|tapacanto|pvc\s*cubrec|cubrecanto|abs|melaminico\s*p/?canto)\b", re.IGNORECASE)

def detectar_tipo(desc):
    for rx, tipo in TIPO_DETECTORS:
        if rx.search(desc):
            return tipo
    return ""

def extraer_espesor(desc):
    m = ESPESOR_RE.search(desc)
    if m:
        return float(m.group(1))
    return ""

def extraer_dims(desc):
    """Devuelve (largo_mm, ancho_mm) o ('', '')."""
    m = DIMS_RE.search(desc)
    if m:
        return (int(float(m.group(1)) * 1000), int(float(m.group(2)) * 1000))
    m = DIMS_ALT_RE.search(desc)
    if m:
        a, b = int(m.group(1)), int(m.group(2))
        # Solo aceptar si parecen dimensiones de placa (>= 1000mm)
        if a >= 1000 and b >= 1000:
            return (a, b)
    return ("", "")

def limpiar_nombre(desc, tipo):
    """Quita los descriptores tecnicos para dejar solo el nombre/color."""
    s = desc
    s = ESPESOR_RE.sub("", s)
    s = DIMS_RE.sub("", s)
    s = DIMS_ALT_RE.sub("", s)
    # Quitar prefijos tipo "Ench Euca MDF", "Mel", etc.
    for rx, _ in TIPO_DETECTORS:
        s = rx.sub("", s)
    # Limpiar abrevs/palabras tecnicas
    s = re.sub(r"\b(euca|euda|rec\s*\w+|simil|ln|x\d{2,}|lc|lex?\d+|xa\d+)\b", "", s, flags=re.IGNORECASE)
    s = re.sub(r"\s+", " ", s).strip(" -·.,")
    return s

def es_placa(desc):
    """Heuristic: tiene tipo detectado AND espesor AND alguna dim."""
    if detectar_tipo(desc) == "":
        return False
    if extraer_espesor(desc) == "":
        return False
    dims = extraer_dims(desc)
    if dims == ("", ""):
        return False
    return True

def es_tapacanto(desc):
    return bool(TAPACANTO_RE.search(desc))

# ── Procesamiento ──────────────────────────────────────────────────────────
print(f"Leyendo {INPUT}...")
with open(INPUT, "r", encoding="utf-8") as f:
    productos = json.load(f)

print(f"Total productos: {len(productos)}")

placas = []
tapacantos = []
descartados = 0

for p in productos:
    desc = p.get("descripcion", "")
    codigo_diac = p.get("codigo_diac", "")
    precio = p.get("precio", "")

    if es_tapacanto(desc):
        tapacantos.append({
            "codigo":         "",  # codigo EGGER/interno - completar manual
            "nombre":         desc,  # sin parsear; el usuario decide
            "espesor":        extraer_espesor(desc) or "",
            "precio":         precio,
            "codigoDiac":     codigo_diac,
            "descripcionDiac": desc,
        })
        continue

    if es_placa(desc):
        tipo = detectar_tipo(desc)
        espesor = extraer_espesor(desc)
        largo, ancho = extraer_dims(desc)
        nombre = limpiar_nombre(desc, tipo)
        placas.append({
            "codigo":         "",   # codigo EGGER/interno - completar manual
            "nombre":         nombre,
            "tipo":           tipo,
            "espesor":        espesor,
            "placaLargo":     largo,
            "placaAncho":     ancho,
            "precioPlaca":    precio,
            "codigoDiac":     codigo_diac,
            "descripcionDiac": desc,
        })
        continue

    descartados += 1

print(f"\nResultados:")
print(f"  Placas extraidas:     {len(placas)}")
print(f"  Tapacantos extraidos: {len(tapacantos)}")
print(f"  Descartados (no son placas ni tapacantos): {descartados}")

# ── CSVs ───────────────────────────────────────────────────────────────────
PLACAS_CSV = OUT_DIR / "lista-placas.csv"
TAPACANTOS_CSV = OUT_DIR / "lista-tapacantos.csv"

placas_sorted = sorted(placas, key=lambda x: (x["tipo"], x["espesor"] or 0, x["nombre"]))
tapacantos_sorted = sorted(tapacantos, key=lambda x: x["nombre"])

with open(PLACAS_CSV, "w", encoding="utf-8", newline="") as f:
    w = csv.DictWriter(f, fieldnames=["codigo", "nombre", "tipo", "espesor", "placaLargo", "placaAncho", "precioPlaca", "codigoDiac", "descripcionDiac"])
    w.writeheader()
    w.writerows(placas_sorted)

with open(TAPACANTOS_CSV, "w", encoding="utf-8", newline="") as f:
    w = csv.DictWriter(f, fieldnames=["codigo", "nombre", "espesor", "precio", "codigoDiac", "descripcionDiac"])
    w.writeheader()
    w.writerows(tapacantos_sorted)

print(f"\nArchivos generados:")
print(f"  {PLACAS_CSV}")
print(f"  {TAPACANTOS_CSV}")
